### PR TITLE
Add the VMM and page mapping

### PR DESCRIPTION
**Pull request by DrDeano**
This adds the VMM to kmain_64 along with the ability to map and unmap paged.

This uses level 4 paging, level 5 can be an option in the future.

Paging isn't initialised into the CR3 register yet as this is in the arch init phase.

Part of #169

### CODE REVIEW

File: src/kernel/arch/x86_64/paging.zig [hunk 1]
```diff
@@ -0,0 +1,806 @@
...
+///
+/// Return: bool
+///     The entry has the attribute.
+///
+fn hasAttribute(entry: u64, attr: u64) callconv(.Inline) bool {
+    return entry & attr == attr;
+}
+
+///
+/// Set or clean the attributes from the VMM attributes to the page entry attributes.
```
Comment (SamTebbs33) [inline src/kernel/arch/x86_64/paging.zig]: clean -> clear

File: src/kernel/arch/x86_64/paging.zig [hunk 2]
```diff
@@ -0,0 +1,806 @@
...
+    var new_pt_entry: u64 = 0;
+    setFrame(&new_pt_entry, physical_addr);
+    setAttribute(&new_pt_entry, ENTRY_PRESENT);
+    setAllAttributes(&new_pt_entry, attrs);
+
+    pt.entries[pt_part] = new_pt_entry;
+}
+
+///
+/// Unmap a the address range from start to end from the Page Table.
```
Comment (SamTebbs33) [inline src/kernel/arch/x86_64/paging.zig]: a the -> an

File: src/kernel/arch/x86_64/paging.zig [hunk 3]
```diff
@@ -0,0 +1,806 @@
...
+            const ptr = @intToPtr([*]Table, frame)[0..1];
+            allocator.free(ptr);
+            // Clear everything
+            pd_entry.* = 0;
+        }
+    }
+}
+
+///
+/// Unmap a the address range from start to end from the Page Directory Pointer Table.
```
Comment (SamTebbs33) [inline src/kernel/arch/x86_64/paging.zig]: a the -> an

File: src/kernel/arch/x86_64/paging.zig [hunk 4]
```diff
@@ -0,0 +1,806 @@
...
+        expectEqual(val, old_val & ~attr);
+    }
+
+    expectEqual(val, 0);
+
+    for (attrs) |attr| {
+        val |= attr;
+    }
+    for (attrs) |attr| {
+        expectEqual(hasAttribute(val, attr), true);
```
Comment (SamTebbs33) [inline src/kernel/arch/x86_64/paging.zig]: This could just be `expect(..)`

**Review by SamTebbs33 — CHANGES_REQUESTED** [review]
Nice one. This looks a hell of a lot more complex than 32 bit

**Comment by SamTebbs33** [comment]
A post-review idea I had was putting some of the function that operator on certain structs (`unmap*Entry`, `map*Entry`, `*Attribute` etc.) inside those structs for some nice encapsulation.

**Comment by DrDeano** [comment]
Could do, add it as an issue. Then can create a new PR for that

File: src/kernel/arch/x86_64/paging.zig [hunk 5]
```diff
@@ -763,7 +763,7 @@ test "map unmap" {
         try unmap(0xFFFFFFFF_00000000, 0xFFFFFFFF_10000000, allocator, &page_table);
         try unmap(0xFFFFFFFF_10000000, 0xFFFFFFFF_F0000000, allocator, &page_table);
 
-        // PML4 entry 511 and PDPT entry 508 and 511 won't be freed as they are not covered but the freeing range
+        // PML4 entry 511 and PDPT entry 508 and 511 won't be freed as they are not covered buy the freeing range
```
Comment (SamTebbs33) [inline src/kernel/arch/x86_64/paging.zig]: buy -> by ;)

File: src/kernel/arch/x86_64/paging.zig [hunk 6]
```diff
@@ -785,7 +785,7 @@ test "map unmap" {
         try map(0xFFFFFFFF_00000000, 0xFFFFFFFF_FFFFF000, 0x0, 0xFFFFF000, .{ .kernel = true, .writable = true, .cachable = true }, allocator, &page_table);
         try unmap(0xFFFFFFFF_00000000, 0xFFFFFFFF_FFFFF000, allocator, &page_table);
 
-        // PML4 entry 511 and PDPT entry 508 and 511 won't be freed as they are not covered but the freeing range
+        // PML4 entry 511 and PDPT entry 508 and 511 won't be freed as they are not covered buy the freeing range
```
Comment (SamTebbs33) [inline src/kernel/arch/x86_64/paging.zig]: buy -> by ;)

**Review by SamTebbs33 — APPROVED** [review]
Looking gucci

**Pull request merged**